Vecmath fixes

gzotti opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Please include a summary of the change and which issue is fixed. -->
<!--- Please also include relevant motivation and context. -->
<!--- List any dependencies that are required for this change. -->

This should add the Vector(2|3|4)(i|f|d) conversions to/from string, HTML color strings and QColors to the Vector classes, removing them from StelUtils. It is merely code cleanup.

DO NOT MERGE! It works only on Windows/MSVC!


### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Test Configuration**:
* Operating system: 
  * Windows 10, MSVC: works
  * Ubuntu 19.10 gcc: cannot link, inline functions are not properly inlined.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings --> (tests do, where appropriate)
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

gzotti commented:
Ouch, I just saw it does not build on Ubuntu.

gzotti commented:
It seems gcc has other inlining capabilities than MSVC. Not sure how to proceed.

guillaumechereau commented:
I think you have to instantiate each template constructors for both float and double in the cpp file (or put the code in the hpp file).  It works on my linux machine with gcc 9.2 if I add those lines in VecMath.cpp:

```
 // Obtains a Vec2i/Vec2f/Vec2d from a string with the form "x,y"
 template<class T> Vector2<T>::Vector2(QString s) : Vector2{s.split(",")}{}
+template Vector2<float>::Vector2(QString s);
+template Vector2<double>::Vector2(QString s);
 
 // Obtains a Vec3i/Vec3f/Vec3d from a string with the form "x,y,z"
 template<class T> Vector3<T>::Vector3(QString s) : Vector3{s.split(",")}{}
+template Vector3<float>::Vector3(QString s);
+template Vector3<double>::Vector3(QString s);
 
 // Obtains a Vec4i/Vec4f/Vec4d from a string with the form "x,y,z,w"
 template<class T> Vector4<T>::Vector4(QString s) : Vector4{s.split(",")}{}
+template Vector4<float>::Vector4(QString s);
+template Vector4<double>::Vector4(QString s);
```

gzotti commented:
Yes, thank you. I just came to the same conclusion :-) Builds again on MSVC, will switch over to Linux/GCC now.

gzotti commented:
@guillaumechereau  not sure about your syntax.
I used now:

```
template<> Vec3i::Vector3(QString s) : Vector3{s.split(",")}{}
template<> Vec3f::Vector3(QString s) : Vector3{s.split(",")}{}
template<> Vec3d::Vector3(QString s) : Vector3{s.split(",")}{}

```
I really dislike this repetition, and had thought the templating would resolve that:

`template<class T> Vector3<T>::Vector3(QString s) : Vector3{s.split(",")}{} // Does not work (?)`

Do your lines really work verbatim? Not for me.

Ah, without the plus in front. That means, I should add the template which includes the initializing code, plus the three declarative lines (for Vec3i, Vec3f, Vec3d) to force generation of the actual code?

guillaumechereau commented:
Without the + yes, that was just there to show which lines to add.

On Wed, Apr 22, 2020 at 11:02 PM Alexander Wolf <notifications@github.com>
wrote:

> *@alex-w* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Stellarium/stellarium/pull/1055#pullrequestreview-398279696>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA2JH7NWFT7IWAF2ZSZSVDRN4AVZANCNFSM4ML565TA>
> .
>
-- 
Sent from my iPad.
